File: 034)ResultReceiverService/app/src/main/java/com/elyesasimsek/resultreceiverservice/MainActivity.java
package com.elyesasimsek.resultreceiverservice;

import android.content.Intent;
import android.os.Bundle;
import android.os.Handler;
import android.os.ResultReceiver;
import android.view.View;
import android.widget.EditText;
import android.widget.TextView;

import androidx.activity.EdgeToEdge;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;

public class MainActivity extends AppCompatActivity {

    private EditText editText;
    private TextView textView;

    Handler handler = new Handler();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        EdgeToEdge.enable(this);
        setContentView(R.layout.activity_main);
        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main), (v, insets) -> {
            Insets systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars());
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom);
            return insets;
        });

        editText = findViewById(R.id.editText);
        textView = findViewById(R.id.textView);
    }

    public void download(View view){
        ResultReceiver myResultReceiver = new MyResultReceiver(null);



        Intent intent = new Intent(this, DownloadClass.class);
        String userInput = editText.getText().toString() ;
        intent.putExtra("url", userInput);
        intent.putExtra("receiver", myResultReceiver);
        startService(intent);
    }

    public class  MyResultReceiver extends ResultReceiver{
        public MyResultReceiver(Handler handler) {
            super(handler);
        }

        @Override
        protected void onReceiveResult(int resultCode, Bundle resultData) {
            super.onReceiveResult(resultCode, resultData);
            if (resultCode == 1 && resultData != null){
                handler.post(new Runnable() {
                    @Override
                    public void run() {
                        String result = resultData.getString("websiteResult");
                        textView.setText(result);
                    }
                });
            }
        }
    }
}

File: 063)CoroutineException/app/src/main/java/com/elyesasimsek/coroutineexception/MainActivity.kt
package com.elyesasimsek.coroutineexception

import android.os.Bundle
import androidx.activity.enableEdgeToEdge
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.lifecycle.lifecycleScope
import kotlinx.coroutines.CoroutineExceptionHandler
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.supervisorScope

class MainActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        enableEdgeToEdge()
        setContentView(R.layout.activity_main)
        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main)) { v, insets ->
            val systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars())
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom)
            insets
        }

        val handler = CoroutineExceptionHandler{ coroutineContext, throwable ->
            println("exception: ${throwable.localizedMessage}")
        }

        lifecycleScope.launch(handler) {
            throw Exception("error")
        }

        lifecycleScope.launch(handler) {
            launch {
                throw Exception("error")
            }

            launch {
                delay(500L)
                println("this is executed")
            }
        }

        lifecycleScope.launch(handler) {
            supervisorScope {
                launch {
                    throw Exception("error")
                }

                launch {
                    delay(500L)
                    println("this is executed")
                }
            }

        }
    }
}